micrometer-test/src/main/java/io/micrometer/core/tck/MeterRegistryAssert.java

```diff
@@ -371,4 +377,115 @@ private String allMetrics() {
         return stringBuilder.toString();
     }
 
+    /**
+     * Finds a meter by name and tags and returns a {@link MeterAssert} for further
+     * assertions.
+     * @param meterName name of the meter
+     * @param tags tags to match
+     * @return a {@link MeterAssert} for the found meter
+     * @throws AssertionError if the meter is not found
+     */
+    public MeterAssert<?> meter(String meterName, Tag... tags) {
+        return this.meter(meterName, Arrays.asList(tags));
+    }
+
+    /**
+     * Finds a meter by name and tags and returns a {@link MeterAssert} for further
+     * assertions.
+     * @param meterName name of the meter
+     * @param tags tags to match
+     * @return a {@link MeterAssert} for the found meter
+     * @throws AssertionError if the meter is not found
+     */
+    public MeterAssert<?> meter(String meterName, Iterable<Tag> tags) {
+        hasMeterWithName(meterName);
+        Meter meter = actual.find(meterName).tags(tags).meter();
+
+        Assertions.assertThat(meter).as("Meter with name <%s> and tags <%s>", meterName, tags).isNotNull();
+        return MeterAssert.assertThat(meter);
+    }
+
+    /**
+     * Finds a counter by name and tags and returns a {@link CounterAssert} for further
+     * assertions.
+     * <p>
+     * Example: <pre><code class='java'>
+     * Counter.builder("my.counter")
+     *     .tag("env", "prod")
+     *     .register(registry);
+     *
+     * MeterRegistryAssert.assertThat(registry)
+     *     .counter("my.counter", Tag.of("env", "prod"))
+     *     .hasCount(0);
+     * </code></pre>
+     * @param name name of the counter
+     * @param tags tags to match
+     * @return a {@link CounterAssert} for the found counter
+     * @throws AssertionError if the counter is not found
+     * @see CounterAssert
+     */
+    public CounterAssert counter(String name, Tag... tags) {
+        MeterAssert<?> meter = meter(name, tags).hasType(Meter.Type.COUNTER);
```

Comment:
It's better to check the Java type is an `instanceof` `Counter`. A custom `Meter` might have `Meter.Type` of `COUNTER` without implementing the `Counter` interface. The cast to `Counter` after will then fail. Similar feedback for the other methods added in this class.